### PR TITLE
[Apiserver decouple] Second part

### PR DESCRIPTION
**Pull request by leonwanghui**
**What this PR does / why we need it**:
See [design spec](https://github.com/opensds/design-specs/pull/17) for detailed information.
Please notice that this PR would address `Phase 2` of the design spec.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
```

### CODE REVIEW

**Comment by leonwanghui** [comment]
This patch contains quite big changes with below items:

* Merge controller and dock protobuf definition into model protobuf located in `model` module.
* Move all logic about status machine into controller module, except that some database operations when deleting volume group still remains in dock module.
* Change `CreateAttachmentOpts` and `UpdateAttachmentOpts` definition in protobuf to `CreateVolumeAttachmentOpts` and `UpdateVolumeAttachmentOpts`.
* Make some changes of volume group operations with dock driver interface, to make it more standard.
* Create `pkg/model/proto/response.go` file for Generating `GenericResponse` struct easily.
* Add some status operations in `pkg/db/db.go`.
* Simplify the ControllerGrpcServer construction by removing `pkg/controller/server.go`.
* Update `e2e_test.go` to make e2e test output more elegant.

File: pkg/api/db.go [hunk 1]
```diff
@@ -66,50 +66,34 @@ func CreateVolumeDBEntry(ctx *c.Context, in *model.VolumeSpec) (*model.VolumeSpe
...
 
 	if volume.Status != model.VolumeAvailable {
-		errMsg := "The status of the volume to be extended must be available"
+		errMsg := "The status of the volume to be extended must be available!"
+		log.Error(errMsg)
+		return nil, errors.New(errMsg)
+	}
+	if newSize <= volume.Size {
+		errMsg := fmt.Sprintf("New size for extend must be greater than current size."+
+			"(current: %d GB, extended: %d GB).", volume.Size, newSize)
```
Comment (xing-yang) [inline pkg/api/db.go]: Error message should start with lower case letter and end without a punctuation.

File: pkg/api/db.go [hunk 2]
```diff
@@ -66,50 +66,34 @@ func CreateVolumeDBEntry(ctx *c.Context, in *model.VolumeSpec) (*model.VolumeSpe
...
-	}
 
-	return result, nil
+	in.Status = model.VolumeCreating
+	// Store the volume data into database.
+	return db.C.CreateVolume(ctx, in)
 }
 
-func ExtendVolumeDBEntry(ctx *c.Context, volID string) (*model.VolumeSpec, error) {
+func ExtendVolumeDBEntry(ctx *c.Context, volID string, newSize int64) (*model.VolumeSpec, error) {
```
Comment (xing-yang) [inline pkg/api/db.go]: Why newSize has to be added now?  How did it work before without the newSize?

File: pkg/api/db.go [hunk 3]
```diff
@@ -66,50 +66,34 @@ func CreateVolumeDBEntry(ctx *c.Context, in *model.VolumeSpec) (*model.VolumeSpe
...
+
 	volume.Status = model.VolumeExtending
 	// Store the volume data into database.
-	result, err := db.C.ExtendVolume(ctx, volume)
-	if err != nil {
-		log.Error("When extend volume in db module:", err)
-		return nil, err
-	}
-	return result, nil
+	return db.C.ExtendVolume(ctx, volume)
```
Comment (xing-yang) [inline pkg/api/db.go]: When is volume.size set to the newSize?

File: pkg/api/db.go [hunk 4]
```diff
@@ -177,31 +140,75 @@ func CreateVolumeSnapshotDBEntry(ctx *c.Context, in *model.VolumeSnapshotSpec) (
...
-	}
+	in.Status = model.VolumeSnapCreating
+	return db.C.CreateVolumeSnapshot(ctx, in)
+}
 
-	result, err := db.C.CreateVolumeSnapshot(ctx, snap)
+func CreateReplicationDBEntry(ctx *c.Context, in *model.ReplicationSpec) (*model.ReplicationSpec, error) {
+	pVol, err := db.C.GetVolume(ctx, in.PrimaryVolumeId)
+	if err != nil {
+		log.Error("Get primary volume failed in create volume replication method: ", err)
```
Comment (xing-yang) [inline pkg/api/db.go]: Get -> get

File: pkg/api/db.go [hunk 5]
```diff
@@ -177,31 +140,75 @@ func CreateVolumeSnapshotDBEntry(ctx *c.Context, in *model.VolumeSnapshotSpec) (
...
+		return nil, err
+	}
+	if pVol.Status != model.VolumeAvailable && pVol.Status != model.VolumeInUse {
+		var errMsg = fmt.Errorf("Only the status of primary volume is available or in-use, the replicaiton can be created")
```
Comment (xing-yang) [inline pkg/api/db.go]: Only -> only

File: pkg/api/db.go [hunk 6]
```diff
@@ -177,31 +140,75 @@ func CreateVolumeSnapshotDBEntry(ctx *c.Context, in *model.VolumeSnapshotSpec) (
...
+		log.Error(errMsg)
+		return nil, errMsg
+	}
+	sVol, err := db.C.GetVolume(ctx, in.SecondaryVolumeId)
 	if err != nil {
-		log.Error("Error occurred in dock module when create volume snapshot in db:", err)
+		log.Error("Get secondary volume failed in create volume replication method: ", err)
```
Comment (xing-yang) [inline pkg/api/db.go]: Starting letter in error message should be lower case.  Same in all other places.

File: contrib/drivers/lvm/lvm.go [hunk 7]
```diff
@@ -547,14 +547,14 @@ func (d *Driver) TerminateSnapshotConnection(opt *pb.DeleteSnapshotAttachmentOpt
 
 }
 
-func (d *Driver) CreateVolumeGroup(opt *pb.CreateVolumeGroupOpts, vg *model.VolumeGroupSpec) (*model.VolumeGroupSpec, error) {
-	return nil, &model.NotImplementError{"Method CreateVolumeGroup did not implement."}
+func (d *Driver) CreateVolumeGroup(opt *pb.CreateVolumeGroupOpts) (*model.VolumeGroupSpec, error) {
+	return nil, &model.NotImplementError{"Method CreateVolumeGroup has not been implemented yet"}
```
[the remark this answers is not in the document]
Comment (leonwanghui) [reply contrib/drivers/lvm/lvm.go]: Got it, but there would be too many changes to be done.

[the remark this answers is not in the document]
Comment (xing-yang) [reply contrib/drivers/lvm/lvm.go]: Only the lines that you have already changed.

File: pkg/api/db.go (hunk 2 above)
Comment (leonwanghui) [reply pkg/api/db.go]: As u can see from below, I added some validation logic in this method.

File: pkg/api/db.go (hunk 3 above)
Comment (leonwanghui) [reply pkg/api/db.go]: It's inside `controller.go` where the status machine is implemented.

File: pkg/api/db.go (hunk 2 above)
Comment (leonwanghui) [reply pkg/api/db.go]: As u can see from below, I added some validation logic in this method.

Comment (xing-yang) [reply pkg/api/db.go]: Some comments should be added to explain why volume.size is not changed to the new size.  If you are just validating, but not really changing the size, explain when the size is really changed.

Also should this validation be done earlier in the API layer, before entering ExtendVolumeDBEntry?

File: pkg/api/db.go (hunk 3 above)
Comment (leonwanghui) [reply pkg/api/db.go]: It's inside `controller.go` where it would be modified in the status machine.

Comment (xing-yang) [reply pkg/api/db.go]: Then add a comment to explain that please.

**Comment by xing-yang** [comment]
@leonwanghui Can you address my comments, please?  Also need a rebase.

**Comment by xing-yang** [comment]
@wisererik Please review this.

File: pkg/api/db.go (hunk 2 above)
Comment (leonwanghui) [reply pkg/api/db.go]: Currently all of validation logic (status, dependency, etc) are implemented in ***DBEntry method, so I just kept that style and one of benefits for that is good for unit test.

**Comment by wisererik** [comment]
@leonwanghui please rebase your PR, thanks

**Comment by xing-yang** [comment]
LGTM

**Pull request merged**